examples/release_highlights/plot_release_highlights_1_5_0.py

```diff
@@ -68,45 +70,60 @@
     classifier, cv=5, scoring="balanced_accuracy"
 ).fit(X, y)
 
-print(f"new threshold: {tuned_classifier.best_threshold_:.4f}")
+print(f"Tuned decision threshold: {tuned_classifier.best_threshold_:.4f}")
 print(
-    f"balanced accuracy: {balanced_accuracy_score(y, tuned_classifier.predict(X)):.2f}"
+    f"Balanced accuracy: {balanced_accuracy_score(y, tuned_classifier.predict(X)):.2f}"
 )
 
 # %%
-# :class:`~model_selection.TunedThresholdClassifierCV` also benefits from the
+# Note however, that the balanced accuracy is not necessarily the most
+# meaningful model selection metric for a given application. It often makes
+# sense to optimize the decision threshold directly for a business metric of
+# interest. **Custom business metrics can be defined by assigning different costs
+# to false positives and false negatives or different gains to true positives
+# and true negatives.** Furthermore, those costs and gains can depend on ancilary
+# metadata specific to each individual data point such as the amount of a
+# transaction in a fraud detection system.
```

Comment:
I think optimizing for custom is the single most important use of this new tool. However the text and the code snippet do not make this point strong enough. I think we need to explain a minimum what a custom metric is and why it can often rely on side metadata to give motivation to the reader to follow the link and read until the section that explains cost-sensitive learning with side metadata.

I agree that this will be too complex to put in the code snippet of the release highlight but I thought this paragraph could help as a middle ground.

> are always not meaningful toy examples just for the purpose on simple demonstration.

Doing things sub-optimally in the past should not be a reason to no improve for the future ;)

If the demonstration is not realistic enough, it's hard to assess the value of the new feature and justify spending time and effort to learn more in the doc / examples.

Reply:
>  I think we need to explain a minimum what a custom metric is and why it can often rely on side metadata to give motivation to the reader to follow the link and read until the section that explains cost-sensitive learning with side metadata.

Alright

> Doing things sub-optimally in the past should not be a reason to no improve for the future

I'm not saying it's suboptimal, to me it's good to make extra simple snippets :)

> If the demonstration is not realistic enough, it's hard to assess the value of the new feature and justify spending time and effort to learn more in the doc / examples.

I'm not sure about that. I have the opposite impression. I'd be interested in seing statistics. To me, it drives more attention if it just shows that you can now do something that you couldn't do before in a very simple way. Then if someone wants to see how far he can go, he will dig the docs deeper.

Reply:
Could we please use a meaningful custom metric then? And avoid all the balanced accuracy stuff?

Reply:
Or simply use accuracy. Here for the threshold, it‘s fine as long as it’s not too extreme.

Reply:
> Could we please use a meaningful custom metric then? And avoid all the balanced accuracy stuff?

This is what I want to motivate by giving reasons to look at the examples instead of blindly using an arbitrary metric such has balanced accuracy. Unfortunately, putting the code of the business metrics directly into the release highlight notebook seems way to much.

> Or simply use accuracy. Here for the threshold, it‘s fine as long as it’s not too extreme.

I tried to tune for accuracy and this leads to a very confusing message. In this case the found optimal threshold is very close to 0.5 but the test accuracy is seemingly slightly lower that sticking with the default (likely in the resampling noise).

Here is a PR to this PR with results where I tuned for accuracy, balanced accuracy and F1 score:

https://github.com/ogrisel/scikit-learn/pull/17

Here is a summary of the results:

```
Default decision threshold: 0.5
Accuracy: 0.9523
Balanced accuracy: 0.8369
F1 score: 0.7475

Decision threshold tuned for accuracy: 0.4638
Accuracy: 0.9516
Balanced accuracy: 0.8435
F1 score: 0.7491

Decision threshold tuned for balanced accuracy: 0.1311
Accuracy: 0.9159
Balanced accuracy: 0.9056
F1 score: 0.6818

Decision threshold tuned for F1 score: 0.3731
Accuracy: 0.9499
Balanced accuracy: 0.8645
F1 score: 0.7552
```

So my idea would be to remove the tuning for accuracy as it's making the main message more complex to explain and only tune and report numbers for F1 score and Balanced accuracy and use that as motivation to learn more on defining meaningful business metrics instead of using semi-arbitrary generic metrics like F1 score and balanced accuracy.

What do you think @lorentzenchr and @jeremiedbb?

Reply:
I think I'd rather keep the current proposal from this PR. I don't see how adding a second metric conveys the message better. If we don't want to promote ``balanced_accuracy``, which I understand, we can replace it by ``roc_auc_score`` ? On Tuning the threshold gives a much lower threshold and shows a net improvement of the score, and I think the explanations also apply for this metric.

Reply:
`roc_auc_score` should be invariant to a change in threshold.
I would not mind to use a very short custom metric like:
```python
from sklearn.metrics import confusion_matrix

def custom_score(y, y_pred):
    """Custom score counting false negatives twice as expensive as true positives."""
    tn, fp, fn, tp = confusion_matrix(y, y_pred).ravel()
    return tp - 2 * fn
```

Reply:
I'm okay with that

Reply:
> `roc_auc_score` should be invariant to a change in threshold.

Yes but I do not see how it's related to this discussion. Using a threshold-invariant metric is pretty useless to tune the decision threshold. Arguably we should even raise a warning in that in that case.

> I would not mind to use a very short custom metric like.

Alright. I will update the PR accordingly then.

Reply:
> Yes but I do not see how it's related to this discussion. Using a threshold-invariant metric is pretty useless to tune the decision threshold. Arguably we should even raise a warning in that in that case.

you can ignore, I made a suggestion but quickly realised it was non-sense so I edited my comment :smile:

Reply:
I pushed 09bd60f5f38cc1a3a29e3b0bd19dcb3335039737. Hopefully it will help getting our message through while staying concise enough.

Reply:
I changed the custom function to actually introduce a tradeoff between the two classes otherwise one would get a dummy classifier without realizing it.